### PR TITLE
msg/async/rdma: fix a coredump introduced by PR #18053,

### PR DESCRIPTION
**Pull request by ownedu**
where the iterator is not working properly after erase().

introduced by #18053

Signed-off-by: Yan Lei <yongyou.yl@alibaba-inc.com>

### CODE REVIEW

**Comment by ownedu** [comment]
@yuyuyu101 pls help review this fix; thanks.

File: src/msg/async/rdma/RDMAStack.cc [hunk 1]
```diff
@@ -244,17 +244,20 @@ void RDMADispatcher::polling()
       perf_logger->set(l_msgr_rdma_inflight_tx_chunks, inflight);
       if (num_dead_queue_pair) {
         Mutex::Locker l(lock); // FIXME reuse dead qp because creating one qp costs 1 ms
-        for (auto &i : dead_queue_pairs) {
+        auto it = dead_queue_pairs.begin();
```
Comment (tchaikov) [inline src/msg/async/rdma/RDMAStack.cc]: my bad, i didn't realize that we are removing elements while iterating thru the vector.

Comment (ownedu) [reply src/msg/async/rdma/RDMAStack.cc]: Thanks for your quick review, anyway.

File: src/msg/async/rdma/RDMAStack.cc [hunk 2]
```diff
@@ -244,17 +244,20 @@ void RDMADispatcher::polling()
...
-            continue;
-          ldout(cct, 10) << __func__ << " finally delete qp=" << i << dendl;
-          delete i;
-          auto it = std::find(dead_queue_pairs.begin(), dead_queue_pairs.end(), i);
-          if (it != dead_queue_pairs.end())
-            dead_queue_pairs.erase(it);
-          perf_logger->dec(l_msgr_rdma_active_queue_pair);
-          --num_dead_queue_pair;
+          if (i->get_tx_wr()) {
+            ldout(cct, 10) << __func__ << " bypass qp=" << i << " tx_wr=" << i->get_tx_wr() << dendl;
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAStack.cc]: I suppose improve this to level 20

Comment (ownedu) [reply src/msg/async/rdma/RDMAStack.cc]: ok

**Comment by ownedu** [comment]
@yuyuyu101 Log level is bumped up, pls help double check; thanks.

**Comment by ownedu** [comment]
@tchaikov could you pls help me clarify the needs-qa label? Does that mean this commit will go though some automatic QA tests before merging? Thanks.

**Comment by tchaikov** [comment]
@ownedu see http://docs.ceph.com/docs/master/dev/#integration-tests-aka-ceph-qa-suite

**Comment by ownedu** [comment]
@tchaikov Regarding the needs-qa, got it and thanks; I will try to do the commit combination.

**Comment by ownedu** [comment]
@tchaikov The latest commit failed with "make check", probably by bad connections, could you pls help take a look? And if so, how to redo "make check"?

The following tests FAILED:
	  2 - run-cli-tests (Failed)
Errors while running CTest
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[ceph-pull-requests] $ /bin/sh -xe /tmp/jenkins7902536043534808187.sh
+ sudo reboot
Agent went offline during the build
ERROR: Connection was broken: java.io.EOFException
	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2638)
	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3113)
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:853)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:349)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:48)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)

**Comment by tchaikov** [comment]
> how to redo "make check"

jenkins, retest this please.

**Comment by ownedu** [comment]
Passed; thanks @tchaikov

**Comment by ownedu** [comment]
@yuyuyu101 @tchaikov QA teuthology testing will be started automatically once it is scheduled?

**Comment by tchaikov** [comment]
@ownedu please define "schedule".

**Comment by ownedu** [comment]
@tchaikov I see "Job stats for this page: 2080 queued, 158 failed, 72 waiting, 102 running, 47 dead, 1348 passed, (3807 total)" in http://pulpito.ceph.com, if I understand correctly this PR's QA is queued and waiting for test.

Plz correct me if I misunderstood anything.

**Comment by tchaikov** [comment]
AFAICT, pulpito is not triggered by any label on github.

> is queued

no

> and waiting for test.

yes.

some kind developer will collect the "needs-qa" PRs for testing them in batch, and analyze the test result, then hopefully merge the tested PRs if all goes well. and it does not happen automatically. normally, this takes up to one week.

**Comment by ownedu** [comment]
@tchaikov gotcha; thanks.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-10-12_16:33:11-rados-wip-kefu-testing-2017-10-12-2302-distro-basic-mira/

**Pull request merged**